aio: autolinking fixes

petebacondarwin opened the pull request:
Closes #20466

petebacondarwin committed:
build(aio): better parsing of selectors as aliases for directives/components

mary-poppins commented:
You can preview c4d5b29 at https://pr20512-c4d5b29.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-base-package/services/filterAmbiguousDirectiveAliases.spec.js:
```diff
@@ -0,0 +1,60 @@
+      expect(filterAmbiguousDirectiveAliases(docs, words, 2)).toEqual(filteredDocs);
+    });
+
+    it('should filter out docs that do not match (case-insensitively) their class name to the matching word', () => {
```
This description seems a little off.

gkalpak commented on aio/tools/transforms/angular-base-package/services/filterAmbiguousDirectiveAliases.js:
```diff
@@ -0,0 +1,34 @@
+    if (docs.length > 1) {
+
+      // Capture the first doc's docType
+      const docType = docs[0].docType;
```
What is the reasoning for caring only about the first doc?

petebacondarwin replied:
because we don't try to do any filtering if the docs are not all of the same type.

gkalpak replied:
And what is the reasoning for not trying to do any filtering if the docs are not all of the same type? :grin:

gkalpak commented on aio/tools/transforms/angular-base-package/services/filterAmbiguousDirectiveAliases.js:
```diff
@@ -0,0 +1,34 @@
+
+        if (docs.every(doc =>
+          // if all the matched docs are of the same type (directives or components)
+          doc.docType === docType &&
```
So, if the first one is of type `component` we don't filter if any of the rest is of type `directive`?
Why is that?

petebacondarwin replied:
In the core docs there are not really many components (if any) so I don't think that this really matters either way. I guess we can be more relaxed and allow any collection of directives and components (just nothing else).

gkalpak commented on aio/tools/transforms/angular-base-package/services/filterAmbiguousDirectiveAliases.js:
```diff
@@ -0,0 +1,34 @@
+          // if all the matched docs are of the same type (directives or components)
+          doc.docType === docType &&
+          // and the matching word is in the selector for all of them
+          doc[docType + 'Options'].selector.indexOf(word) != -1
```
Should we take word boundaries into account. E.g. `['ngModelStatus'].indexOf('ngModel') !== -1`, but should that count as a match?

petebacondarwin replied:
I guess you are saying that you think it should not?

petebacondarwin replied:
I wonder if this is likely to be a problem...?
For a `doc` to appear in the original `docs` list that is passed to the filter function, the matching word must appear as an `alias` of the `doc`. So I think that it is unlikely that the matching word does not appear in the selector at all.

petebacondarwin committed:
build(aio): do not store duplicate metadata aliases

Having duplicates was causing the code auto-linking
to ignore `ngForm` directives.

petebacondarwin committed:
build(aio): filter out ambiguous directives from auto code linking

Closes #20466

mary-poppins commented:
You can preview dad6f35 at https://pr20512-dad6f35.ngbuilds.io/.

gkalpak commented:
The Travis failures are unrelated to the PR.

IgorMinar commented:
@petebacondarwin please add target label in the future. in general for docs, bug fixes and infrastructure changes we should always try to cherrypick the commit to the stable branch

gkalpak commented:
Ooops! I added the `merge` label without checking for `target` :flushed:

petebacondarwin commented:
@IgorMinar yes, I normally add the target to my own PRs when I add the merge label. I didn't realise that @gkalpak had added the merge label when he approved it.
